src/_backend_agg_basic_types.h

```diff
@@ -121,4 +126,135 @@ class GCAgg
     GCAgg &operator=(const GCAgg &);
 };
 
+namespace PYBIND11_NAMESPACE { namespace detail {
+    template <> struct type_caster<agg::line_cap_e> {
+    public:
+        PYBIND11_TYPE_CASTER(agg::line_cap_e, const_name("line_cap_e"));
+
+        bool load(handle src, bool) {
+            const std::unordered_map<std::string, agg::line_cap_e> enum_values = {
+                {"butt", agg::butt_cap},
+                {"round", agg::round_cap},
+                {"projecting", agg::square_cap},
+            };
+            value = enum_values.at(src.cast<std::string>());
+            return true;
+        }
+    };
+
+    template <> struct type_caster<agg::line_join_e> {
+    public:
+        PYBIND11_TYPE_CASTER(agg::line_join_e, const_name("line_join_e"));
+
+        bool load(handle src, bool) {
+            const std::unordered_map<std::string, agg::line_join_e> enum_values = {
+                {"miter", agg::miter_join_revert},
+                {"round", agg::round_join},
+                {"bevel", agg::bevel_join},
+            };
+            value = agg::miter_join_revert;
+            value = enum_values.at(src.cast<std::string>());
+            return true;
+        }
+    };
+
+    template <> struct type_caster<ClipPath> {
+    public:
+        PYBIND11_TYPE_CASTER(ClipPath, const_name("ClipPath"));
+
+        bool load(handle src, bool) {
+            if (src.is_none()) {
+                return true;
+            }
+
+            auto clippath_tuple = src.cast<py::tuple>();
```

Comment:
Would probably write as something like (untested...)
```c++
auto [path, trans] = src.cast<std::pair<std::optional<mpl::PathIterator>, agg::trans_affine>>();
if (path) { value.path = *path; }
value.trans = trans;
```
(note that the current implementation doesn't actually check the length of the tuple).
Ditto for dashes below.

Reply:
I thought I had tried this and ran into something weird with `std::optional`, but it seems to work. Maybe it had something to do with parameters directly.